hibernate-core/src/main/java/org/hibernate/query/sqm/tree/expression/SqmBinaryArithmetic.java

```diff
@@ -52,8 +54,20 @@ public SqmBinaryArithmetic(
 				( operator == ADD || operator == SUBTRACT ) ) {
 			return;
 		}
-		this.lhsOperand.applyInferableType( rhsOperand.getNodeType() );
-		this.rhsOperand.applyInferableType( lhsOperand.getNodeType() );
+		this.lhsOperand.applyInferableType( getExpressibleNodeType( rhsOperand ) );
+		this.rhsOperand.applyInferableType( getExpressibleNodeType( lhsOperand ) );
+	}
+
+	private static SqmExpressible<?> getExpressibleNodeType(SqmExpression<?> arg) {
+		if ( arg instanceof SqmPath ) {
+			SqmPath path = (SqmPath) arg;
+			if ( path.getModel() instanceof SqmPathSource ) {
+				if ( ( (SqmPathSource) path.getModel() ).isGeneric() ) {
+					return path.getResolvedModel();
+				}
+			}
+		}
+		return arg.getNodeType();
```

Comment:
Please remove this helper.

Reply:
I was sure that I've removed that :-( Will recheck.